[clang] fix half && bfloat16 convert node expr codegen

JinjinLi868 opened the pull request:
Data type conversion between fp16 and bf16 will generate fptrunc and fpextend nodes, but they are actually bitcast nodes.

arsenm commented on clang/test/CodeGen/X86/bfloat16-convert-half.c:
```diff
@@ -0,0 +1,14 @@
+//
+_Float16 test_convert(__bf16 a) {
+    return (_Float16)a;
+}
```
Can you also test some vector cases?

arsenm reviewed (commented):
This appears to just assert today, but interpreting this as bitcast doesn't make sense. I would expect this to emit a pair of casts, fpext to float, and fptrunc down to half

arsenm commented:
> This appears to just assert today, but interpreting this as bitcast doesn't make sense. I would expect this to emit a pair of casts, fpext to float, and fptrunc down to half

If we don't just reject it as an invalid cast

JinjinLi868 commented:
> This appears to just assert today, but interpreting this as bitcast doesn't make sense. I would expect this to emit a pair of casts, fpext to float, and fptrunc down to half

JinjinLi868 commented:
> > This appears to just assert today, but interpreting this as bitcast doesn't make sense. I would expect this to emit a pair of casts, fpext to float, and fptrunc down to half
> 
> If we don't just reject it as an invalid cast

i understand your means, it like a Hardware behavior to do fp16 convert to bf16（firstly, fp16 fpext to float32, second , float32 fptrunc to bfloat16).  But In some target, it supply a HW instruction to complete the process  (fp16->float32->bf16) . so it just supply a intrinsic (fp16 -> bf16)

arsenm commented on clang/test/CodeGen/X86/bfloat16-convert-half.c:
```diff
@@ -0,0 +1,109 @@
+// CHECK-NEXT:    ret i32 [[TMP1]]
+//
+__bf162 test_cast_from_fp162_to_bf162(_Float162 in) {
+  return __builtin_convertvector(in, __bf162);
```
I meant test the raw cast, not this builtin

arsenm commented on clang/test/CodeGen/X86/bfloat16-convert-half.c:
```diff
@@ -0,0 +1,109 @@
+    return (__bf16)a;
+}
+
+typedef _Float16 _Float162 __attribute__((ext_vector_type(2)));
+typedef _Float16 _Float164 __attribute__((ext_vector_type(4)));
+
+typedef __bf16 __bf162 __attribute__((ext_vector_type(2)));
+typedef __bf16 __bf164 __attribute__((ext_vector_type(4)));
```
Can you use a different typedef name? Maybe half2/bfloat2?

arsenm reviewed (changes requested):
Bitcast is not the correct behavior

arsenm commented:
> But In some target, it supply a HW instruction to complete the process (fp16->float32->bf16) . so it just supply a intrinsic (fp16 -> bf16)

Which is not a bitcast. The correct IR representation of this conversion is fpext+fptrunc

JinjinLi868 commented:
> > But In some target, it supply a HW instruction to complete the process (fp16->float32->bf16) . so it just supply a intrinsic (fp16 -> bf16)
> 
> Which is not a bitcast. The correct IR representation of this conversion is fpext+fptrunc

i understand, i have changed. but for the vector convert case, it is also a bitcast IR

efriedma-quic commented:
For the vector case, I think you want __builtin_convertvector or something like that

JinjinLi868 commented:
> For the vector case, I think you want __builtin_convertvector or something like that

thanks, i have add  __builtin_convertvector testcase

efriedma-quic commented on clang/test/CodeGen/X86/bfloat16-convert-half.c:
```diff
@@ -0,0 +1,194 @@
+// CHECK-NEXT:    [[IN1:%.*]] = load <2 x half>, ptr [[IN]], align 4
+// CHECK-NEXT:    store <2 x half> [[IN1]], ptr [[IN_ADDR]], align 4
+// CHECK-NEXT:    [[TMP0:%.*]] = load <2 x half>, ptr [[IN_ADDR]], align 4
+// CHECK-NEXT:    [[TMP1:%.*]] = bitcast <2 x half> [[TMP0]] to <2 x bfloat>
```
The CastKind in the AST is BitCast.  This is inherited from old GNU vector_type semantics.  Whether or not that's correct is orthogonal to this patch.

arsenm replied:
Does GCC have the same behavior for the bfloat x half case?

efriedma-quic reviewed (approved):
LGTM

wangpc-pp commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1431,9 +1431,13 @@ Value *ScalarExprEmitter::EmitScalarCast(Value *Src, QualType SrcType,
 
-  if (DstElementTy->getTypeID() < SrcElementTy->getTypeID())
+  if ((DstElementTy->is16bitFPTy() && SrcElementTy->is16bitFPTy())) {
+    Value *FloatVal = Builder.CreateFPExt(Src, Builder.getFloatTy(), "conv");
+    return Builder.CreateFPTrunc(FloatVal, DstTy, "conv");
```
```suggestion
    Value *FloatVal = Builder.CreateFPExt(Src, Builder.getFloatTy(), "fpext");
    return Builder.CreateFPTrunc(FloatVal, DstTy, "fptunc");
```
This should make the IR more readable.

wangpc-pp commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1906,7 +1910,15 @@ Value *ScalarExprEmitter::VisitConvertVectorExpr(ConvertVectorExpr *E) {
+          llvm::VectorType::get(Builder.getFloatTy(),
+                                ScrVecTy->getElementCount()),
+          "conv");
+      Res = Builder.CreateFPTrunc(FloatVal, DstTy, "conv");
```
ditto.

wangpc-pp commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1906,7 +1910,15 @@ Value *ScalarExprEmitter::VisitConvertVectorExpr(ConvertVectorExpr *E) {
+          Src,
+          llvm::VectorType::get(Builder.getFloatTy(),
+                                ScrVecTy->getElementCount()),
+          "conv");
```
ditto.

wangpc-pp commented on clang/test/CodeGen/X86/bfloat16-convert-half.c:
```diff
@@ -0,0 +1,194 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 4
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -target-feature +fullbf16 -S -emit-llvm %s -o - | FileCheck %s
```
```suggestion
// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -target-feature +fullbf16 -disable-O0-optnone -emit-llvm %s -o - | opt -S -passes=mem2reg | FileCheck %s
```
We can run `mem2reg` to reduce CHECKs.

wangpc-pp commented on clang/test/CodeGen/X86/bfloat16-convert-half.c:
```diff
@@ -0,0 +1,194 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 4
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -target-feature +fullbf16 -S -emit-llvm %s -o - | FileCheck %s
+// CHECK-LABEL: define dso_local half @test_convert_from_bf16_to_fp16(
```
There should be a blank between `RUN` and `CHECK`.

arsenm commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1431,9 +1431,13 @@ Value *ScalarExprEmitter::EmitScalarCast(Value *Src, QualType SrcType,
+  if ((DstElementTy->is16bitFPTy() && SrcElementTy->is16bitFPTy())) {
+    Value *FloatVal = Builder.CreateFPExt(Src, Builder.getFloatTy(), "conv");
+    return Builder.CreateFPTrunc(FloatVal, DstTy, "conv");
+  } else if (DstElementTy->getTypeID() < SrcElementTy->getTypeID())
```
No else after return (although it's already violated in the existing code)

topperc replied:
Why was this resolved? It doesn't seem to have been addressed.

arsenm replied:
Still needs to be resolved. Also should probably get a comment

arsenm reviewed (approved):
LGTM. Would be good to verify the vector case is "correct" in as far as it's what GCC does

arsenm commented on clang/test/CodeGen/X86/bfloat16-convert-half.c:
```diff
@@ -0,0 +1,165 @@
+  return __builtin_convertvector(in, half2);
+}
+
+// CHECK-LABEL: define dso_local double @test_convertvector_from_half4_to_bfloat4(
```
Is the double here some weird ABI thing or intentional?

wangpc-pp replied:
It follows X86's ABI I think.

arsenm commented on clang/test/CodeGen/X86/bfloat16-convert-half.c:
```diff
@@ -0,0 +1,25 @@
+__bf16 test_convert_from_fp16_to_bf16(_Float16 a) {
+    return (__bf16)a;
+}
+
```
The regular C case should still test the vector conversions. Also the rebase is confusing me

wangpc-pp replied:
Vector tests are moved to HIP target now as the IRs are wierd in X86.

arsenm replied:
I think these tests need to be additive. The vector behavior seems to be different between standard C and the proper vector languages?

arsenm commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1906,7 +1909,15 @@ Value *ScalarExprEmitter::VisitConvertVectorExpr(ConvertVectorExpr *E) {
     assert(SrcEltTy->isFloatingPointTy() && DstEltTy->isFloatingPointTy() &&
            "Unknown real conversion");
-    if (DstEltTy->getTypeID() < SrcEltTy->getTypeID())
+    if ((DstEltTy->is16bitFPTy() && SrcEltTy->is16bitFPTy())) {
```
This should be a separate change if it's correct

JinjinLi868 commented:
ping

arsenm commented:
> ping Ping Do you have another review comment?

This has now confused me. You should roll back to the case where you only changed the scalar behavior. Any vector behavior change should be a separate PR, if that is even correct. I would still like to know what the gcc behavior is in this case

topperc commented:
@JinjinLi868 are you still working on this?

wangpc-pp commented:
> @JinjinLi868 are you still working on this?

I can ask him. Is this PR blocking some of your works on float16/bf16?

JinjinLi868 commented:
> @JinjinLi868 are you still working on this?

yes, i am. could you give me some advice and can i help you ?

topperc commented:
> > @JinjinLi868 are you still working on this?
> 
> I can ask him. Is this PR blocking some of your recent works on float16/bf16?

I stumbled onto the verifier error earlier while writing a test. It's not blocking me.

topperc commented:
> > @JinjinLi868 are you still working on this?
> 
> yes, i am. could you give me some advice and can i help you ?

Can we have a scalar only patch as @arsenm requested?

JinjinLi868 commented:
> > > @JinjinLi868 are you still working on this?
> > 
> > 
> > yes, i am. could you give me some advice and can i help you ?
> 
> Can we have a scalar only patch as @arsenm requested?

ok, you mean, i remove the vector testcase for this patch. and just save the scalar testcase?

arsenm commented:
> ok, you mean, i remove the vector testcase for this patch. and just save the scalar testcase?

No, keep the tests. Only keep the scalar behavior change. The previous revision was essentially correct and minimal

JinjinLi868 committed:
[clang] Fix half && bfloat16 convert node expr codegen

Data type conversion between fp16 and bf16 will generate fptrunc
and fpextend nodes, but they are actually bitcast nodes.

JinjinLi868 commented:
> > > > @JinjinLi868 are you still working on this?
> > > 
> > > 
> > > yes, i am. could you give me some advice and can i help you ?
> > 
> > 
> > Can we have a scalar only patch as @arsenm requested?
> 
> ok, you mean, i remove the vector testcase for this patch. and just save the scalar testcase?

i have changed. just save the scalar on this patch

JinjinLi868 commented:
> > ok, you mean, i remove the vector testcase for this patch. and just save the scalar testcase?
> 
> No, keep the tests. Only keep the scalar behavior change. The previous revision was essentially correct and minimal

i have changed. just save the scalar on this patch

arsenm reviewed (commented):
The vector tests should still be added

JinjinLi868 commented:
> The vector tests should still be added

sorry. if i remove the change of the vector. i have to remove the testcase. because, for the current code convert between vector type of half and bfloat16, it has a bug. And it will be Assert "Invalid cast!""
`CastInst *CastInst::Create(Instruction::CastOps op, Value *S, Type *Ty,
                           const Twine &Name, InsertPosition InsertBefore) {
  assert(castIsValid(op, S, Ty) && "Invalid cast!");
  // Construct and return the appropriate CastInst subclass
  switch (op) {
  case Trunc:         return new TruncInst         (S, Ty, Name, InsertBefore);
  case ZExt:          return new ZExtInst          (S, Ty, Name, InsertBefore);
  case SExt:          return new SExtInst          (S, Ty, Name, InsertBefore);
  case FPTrunc:       return new FPTruncInst       (S, Ty, Name, InsertBefore);
  case FPExt:         return new FPExtInst         (S, Ty, Name, InsertBefore);

  default:
    llvm_unreachable("Invalid opcode provided");
  }
}`
`CastInst::castIsValid(Instruction::CastOps op, Type *SrcTy, Type *DstTy) {
  if (!SrcTy->isFirstClassType() || !DstTy->isFirstClassType() ||
      SrcTy->isAggregateType() || DstTy->isAggregateType())
    return false;

  // Get the size of the types in bits, and whether we are dealing
  // with vector types, we'll need this later.
  bool SrcIsVec = isa<VectorType>(SrcTy);
  bool DstIsVec = isa<VectorType>(DstTy);
  unsigned SrcScalarBitSize = SrcTy->getScalarSizeInBits();
  unsigned DstScalarBitSize = DstTy->getScalarSizeInBits();

  // If these are vector types, get the lengths of the vectors (using zero for
  // scalar types means that checking that vector lengths match also checks that
  // scalars are not being converted to vectors or vectors to scalars).
  ElementCount SrcEC = SrcIsVec ? cast<VectorType>(SrcTy)->getElementCount()
                                : ElementCount::getFixed(0);
  ElementCount DstEC = DstIsVec ? cast<VectorType>(DstTy)->getElementCount()
                                : ElementCount::getFixed(0);

  // Switch on the opcode provided
  switch (op) {
  default: return false; // This is an input error
  case Instruction::FPExt:
    return SrcTy->isFPOrFPVectorTy() && DstTy->isFPOrFPVectorTy() &&
           SrcEC == DstEC && SrcScalarBitSize < DstScalarBitSize;
`
now, for the vector convert between half and bfloat16. it codegen to the FPExt. For the FPExt,  castIsValid() need SrcScalarBitSize < DstScalarBitSize;  But for vector half and bfloat16, the SrcScalarBitSize is equal to DstScalarBitSize.

JinjinLi868 commented:
i add the gcc behavior in this case 
```
__bf16 test_convert_from_fp16_to_bf16(_Float16 a) {
    return (__bf16)a;
}
test_convert_from_fp16_to_bf16(_Float16):
        push    rbp
        mov     rbp, rsp
        sub     rsp, 16
        movd    eax, xmm0
        mov     WORD PTR [rbp-2], ax
        pinsrw  xmm0, WORD PTR [rbp-2], 0
        call    __trunchfbf2
        movd    eax, xmm0
        movd    xmm0, eax
        leave
        ret
```
```
_Float16 test_convert_from_bf16_to_fp16(__bf16 a) {
    return (_Float16)a;
}
test_convert_from_bf16_to_fp16(std::bfloat16_t):
        push    rbp
        mov     rbp, rsp
        sub     rsp, 16
        movd    eax, xmm0
        mov     WORD PTR [rbp-2], ax
        movzx   eax, WORD PTR [rbp-2]
        sal     eax, 16
        movd    xmm0, eax
        call    __truncsfhf2
        movd    eax, xmm0
        movd    xmm0, eax
        leave
        ret
```

arsenm commented:
> > The vector tests should still be added
> 
> sorry. if i remove the change of the vector. i have to remove the testcase. because, for the current code convert between vector type of half and bfloat16, it has a bug. And it will be Assert "Invalid cast!""
> 

OK, LGTM with the else before return fixed. Can you handle the vector case in a follow up?

JinjinLi868 commented:
> > > The vector tests should still be added
> > 
> > 
> > sorry. if i remove the change of the vector. i have to remove the testcase. because, for the current code convert between vector type of half and bfloat16, it has a bug. And it will be Assert "Invalid cast!""
> 
> OK, LGTM with the else before return fixed. Can you handle the vector case in a follow up?

OK, i will handle the vector case in a follow up.
